### PR TITLE
Rearrange training completely, using browser view on the context

### PR DESCRIPTION
**Pull request by pysailor**

### CODE REVIEW

File: src/euphorie/client/browser/training.py [hunk 1]
```diff
@@ -12,43 +19,207 @@ class TrainingSlide(BrowserView):
...
+
+    @property
+    @memoize
+    def existing_measures(self):
+        try:
+            existing_measures = loads(self.context.existing_measures)
+            # Backwards compat. We used to save dicts in JSON before we
+            # switched to list of tuples.
+            if isinstance(existing_measures, dict):
+                existing_measures = [(k, v) for (k, v) in existing_measures.items()]
```
Comment (ale-rt) [inline src/euphorie/client/browser/training.py]:
```
existing_measures = existing_measures.items()
```
is the same as:
```
existing_measures = [(k, v) for (k, v) in existing_measures.items()]

File: src/euphorie/client/browser/training.py [hunk 2]
```diff
@@ -12,43 +19,207 @@ class TrainingSlide(BrowserView):
...
+            existing_measures = [
+                text.strip() for (text, active) in existing_measures if active
+            ]
+        except:
+            existing_measures = []
```
Comment (ale-rt) [inline src/euphorie/client/browser/training.py]: I would not catch a bare exception without at least logging something...
can we add a:
```
logger.exception("Something is broken in %r", self.context)
```

File: src/euphorie/client/browser/training.py [hunk 3]
```diff
@@ -12,43 +19,207 @@ class TrainingSlide(BrowserView):
...
+        return existing_measures
+
+    @property
+    @memoize
+    def image(self):
+        if self.is_custom:
+            if getattr(self.context, "image_data", None):
```
Comment (ale-rt) [inline src/euphorie/client/browser/training.py]: Anti if:
```
if getattr(self.context, "image_data", None):
    return
```

File: src/euphorie/client/browser/training.py [hunk 4]
```diff
@@ -12,43 +19,207 @@ class TrainingSlide(BrowserView):
...
+                _view = self.context.__of__(
```
Comment (ale-rt) [inline src/euphorie/client/browser/training.py]: It seems strange to me that this is needed

File: src/euphorie/client/browser/training.py [hunk 5]
```diff
@@ -12,43 +19,207 @@ class TrainingSlide(BrowserView):
...
+                    self.webhelpers.traversed_session.aq_parent["custom-risks"]
+                ).restrictedTraverse("image-display")
+                _view.field_name = "image_large"
+                return _view.get_or_create_image_scaled()
+            return None
+        image = self.zodb_elem.image and self.zodb_elem.image.data or None
```
Comment (ale-rt) [inline src/euphorie/client/browser/training.py]: It seems to me there is no point in getting the image if `self.for_download` is False.

File: src/euphorie/client/browser/training.py [hunk 6]
```diff
@@ -12,43 +19,207 @@ class TrainingSlide(BrowserView):
...
+        if image and self.for_download:
+            try:
+                scales = self.zodb_elem.restrictedTraverse("images", None)
+                if self.item_type == "module":
```
Comment (ale-rt) [inline src/euphorie/client/browser/training.py]: What is the point of computing the scale_name if scales is falsish?

File: src/euphorie/client/browser/training.py [hunk 7]
```diff
@@ -12,43 +19,207 @@ class TrainingSlide(BrowserView):
...
+                    scale_name = "training_title"
+                else:
+                    scale_name = "training"
+                if scales:
+                    scale = scales.scale(fieldname="image", scale=scale_name)
+                    if scale and scale.data:
+                        image = scale.data.data
+            except:
+                image = None
```
Comment (ale-rt) [inline src/euphorie/client/browser/training.py]: I would not catch a bare exception without at least logging something...

File: src/euphorie/client/browser/training.py [hunk 8]
```diff
@@ -12,43 +19,207 @@ class TrainingSlide(BrowserView):
...
+        return "for_download" in self.request
+
+    @property
+    @memoize
+    def title_image(self):
+        survey = self.context.aq_parent
+        return (
+            getattr(survey, "external_site_logo", None)
+            and survey.external_site_logo.data
+            or None
```
Comment (ale-rt) [inline src/euphorie/client/browser/training.py]: It seems a case were a try/except should be preferred to a getattr
```
try:
     return self.context.aq_parent.external_site_logo.data
except AttributeError:
     return
```

File: src/euphorie/client/browser/training.py [hunk 9]
```diff
@@ -12,43 +19,207 @@ class TrainingSlide(BrowserView):
...
+        )
+
+    @property
+    @memoize
+    def slide_data(self):
+        modules = self.getModulePaths()
+        risks = self.getRisks(modules, skip_unanswered=self.skip_unanswered)
```
Comment (ale-rt) [inline src/euphorie/client/browser/training.py]: Looks strange to pass an instance attribute (`self.skip_unanswered`) to an instance method.

File: src/euphorie/client/browser/training.py (hunk 3 above)
Comment (pysailor) [reply src/euphorie/client/browser/training.py]: I think you mean `if not getattr(self.context, "image_data", None)` ?

Comment (ale-rt) [reply src/euphorie/client/browser/training.py]: yes, not ;)

**Comment by pysailor** [comment]
Thanks for the review, @ale-rt ! I made the suggested simplifications. I'll merge once the tests are green again.

**Pull request merged**